source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -476,41 +476,112 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
+                                                     : XID_LENGTH + ZXID_LENGTH + INT_LENGTH);
       ensureMaxLength(len);
       offset += len;
+      if (offset <= data_len) {
+        has_full_packets = true;
+      }
     } catch (const EnvoyException& e) {
       ENVOY_LOG(debug, "zookeeper_proxy: decoding exception {}", e.what());
       callbacks_.onDecodeError();
-      return Network::FilterStatus::Continue;
+      return;
     }
   }
 
-  if (offset == data.length()) {
+  if (offset == data_len) {
     decode(data, dtype);
-    return Network::FilterStatus::Continue;
-  }
 
-  ENVOY_LOG(trace,
-            "zookeeper_proxy: waiting for entire packets in the buffer, current buffer "
-            "is {} bytes",
-            data.length());
-  return Network::FilterStatus::StopIteration;
+    if (zk_filter_buffer_len > 0) {
+      // After prepending the ZooKeeper filter buffer to the network filter buffer, the network
+      // filter buffer here includes one or more full packets without any partial data, the decoder
+      // is able to decode the entire network filter buffer. So the we need to drain the ZooKeeper
+      // filter buffer here, since its data has been decoded already. Otherwise, the ZooKeeper
+      // filter will prepend the ZooKeeper filter buffer to the network filter buffer again next
+      // time, which is not what we want.
+      zk_filter_buffer.drain(zk_filter_buffer_len);
+    }
+  } else if (offset > data_len) {
+    std::string temp_data;
+
+    if (has_full_packets) {
+      offset -= INT_LENGTH + len;
+      // Decode full packets.
+      temp_data.resize(offset);
+      data.copyOut(0, offset, temp_data.data());
+      Buffer::OwnedImpl full_packets;
+      full_packets.add(temp_data.data(), temp_data.length());
+      decode(full_packets, dtype);
+      // After prepending the ZooKeeper filter buffer to the network filter buffer, the network
+      // filter buffer here includes one or more full packets with partial data of a packet, the
+      // decoder is able to decode all full packets. So we need to drain the ZooKeeper filter buffer
+      // here, since its data has been decoded already. Otherwise, the ZooKeeper filter will prepend
+      // the ZooKeeper filter buffer to the network filter buffer again next time, which is not what
+      // we want.
+      zk_filter_buffer.drain(zk_filter_buffer_len);
+
+      // Copy out the rest of the data to the ZooKeeper filter buffer.
+      temp_data.resize(data_len - offset);
+      data.copyOut(offset, data_len - offset, temp_data.data());
+      zk_filter_buffer.add(temp_data.data(), temp_data.length());
+    } else {
+      // Copy out all the new data to the ZooKeeper filter buffer.
+      temp_data.resize(data_len - zk_filter_buffer_len);
+      data.copyOut(zk_filter_buffer_len, data_len - zk_filter_buffer_len, temp_data.data());
+      zk_filter_buffer.add(temp_data.data(), temp_data.length());
```

Comment:
it's awful to copy twice but I can't find an easy way to save a copy except bufferFragment :(